build/scripts/patch-prepare-repo.sh

```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+# Copyright 2024 The Cockroach Authors.
+#
+# Use of this software is governed by the CockroachDB Software License
+# included in the /LICENSE file.
+
+set -euo pipefail
+
+function usage() {
+  echo "This script prepares a clone of a dependency for applying and generating"
+  echo "patches. It generates a WORKSPACE file and runs gazelle to generate"
+  echo "BUILD.bazel files and commits these changes."
+  echo ""
+  echo "This script should be used to set up the \"base\" tree before using"
+  echo "patch-apply.sh and patch-gen.sh."
+  echo ""
+  echo "Usage: $0 <bazel-repo-name> <repo-directory> [flags for gazelle]"
+  echo ""
+  echo "Example: $0 com_github_cockroachdb_pebble ~/pebble"
+  exit 1
+}
+
+# Check for required arguments
+if [ "$#" -lt 2 ]; then
+  usage
+fi
+
+REPO_NAME="$1"
+REPO_DIR="$2"
+
+# Generate the import path from the workspace. For example,
+# com_github_cockroachdb_pebble.patch to github.com/cockroachdb/pebble.
+
+# Split the input string by underscores
+IFS='_' read -r -a fields <<< "$REPO_NAME"
+
+# Extract the first two fields
+field1="${fields[0]}"
+field2="${fields[1]}"
+
+# Join the remaining fields with '/'
+rest=$(IFS=/; echo "${fields[*]:2}")
+
+IMPORT_PATH="${field2}.${field1}/${rest}"
+
+echo "Import path: $IMPORT_PATH"
+
+# Check if the repository directory exists.
+if [ ! -d "$REPO_DIR" ]; then
+  echo "Error: Repository directory '$REPO_DIR' does not exist."
+  exit 1
+fi
+
+cd "$REPO_DIR"
+
+if grep -q -E '^[::space::]*(/WORKSPACE|BUILD.bazel)[::space::]*$' .gitignore; then
+  echo "Removing /WORKSPACE and BUILD.bazel from .gitignore"
+
+  TEMP_FILE=$(mktemp)
+  cp .gitignore $TEMP_FILE
+  sed -E '\#^[::space::]*(/WORKSPACE|BUILD.bazel)[::space::]*$#d' $TEMP_FILE >.gitignore
+
+  git commit .gitignore -m "gitignore: /WORKSPACE and BUILD.bazel"
+fi
+
+if [ ! -f WORKSPACE ]; then
+	echo "Generating WORKSPACE"
+	echo "workspace(name = \"$REPO_NAME\")" > WORKSPACE
+fi
+
+echo "Running gazelle update ${@:3} --go_prefix=\"$IMPORT_PATH\" --repo_root=."
+go run github.com/bazelbuild/bazel-gazelle/cmd/gazelle@v0.37.0 update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=.
```

Comment:
Did you try using `bazel run @bazel_gazelle//cmd/gazelle --run_under="cd $PWD && "` (or something to that effect)? This would mean that we don't have to keep this script in sync with the version of `gazelle` used in `WORKSPACE`.

Reply:
Good idea, I'll try it.

Reply:
I'm having some issues making it work. The command succeeds but it does nothing.

I'm doing
```
  bazel run @bazel_gazelle//cmd/gazelle --run_under="cd $REPO_DIR && " -- update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=.
```

Output:
```
+ bazel run @bazel_gazelle//cmd/gazelle '--run_under=cd /Users/radu/pebble && ' -- update --go_prefix=github.com/cockroachdb/pebble --repo_root=.
INFO: Invocation ID: d001ebae-f750-4880-bc4c-18010a3bcd26
INFO: Analyzed target @@bazel_gazelle//cmd/gazelle:gazelle (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target @@bazel_gazelle//cmd/gazelle:gazelle up-to-date:
  _bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle
INFO: Elapsed time: 0.166s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: /bin/bash -c 'cd /Users/radu/pebble &&  _bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle update '\''--go_prefix=github.com/cockroachdb/pebble'\'' '\''--repo_root=.'\'''
```

This looks like it wouldn't work because `_bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle` is no longer valid after `cd`.

I see `run_under` is used similarly in other scripts, not sure why it doesn't work in this case.

I couldn't find a way to make gazelle run correctly with the crdb root as the cwd.

Reply:
It does run the command (it shows the help if I pass `-help` at the end). But it just doesn't do anything, still trying to figure out why.

Reply:
Not sure what's with the weird escaping of the update arguments. It works if I manually run (from the other repo)
```
~/roach/_bazel/bin/external/bazel_gazelle/cmd/gazelle/gazelle_/gazelle update --go_prefix=github.com/cockroachdb/pebble --repo_root=.
```

Reply:
Still not sure why the run_under stuff didn't work but I made it work like this:
```
bazel run @bazel_gazelle//cmd/gazelle -- update ${@:3} --go_prefix="$IMPORT_PATH" --repo_root=$REPO_DIR $REPO_DIR
```

Reply:
Nice.